【edk2和opensbi的配套RAS功能已提出pr】riscv: add RAS support with SBI SSE extension

GooTal opened the pull request:
fixed: #95

community inclusion
category: feature
bugzilla: https://github.com/RVCK-Project/rvck/issues/95
Link: https://lore.kernel.org/all/20250908181717.1997461-2-cleger@rivosinc.com/
Link: https://lore.kernel.org/all/20250227123628.2931490-2-hchauhan@ventanamicro.com/

-------------------------------------------------

Implement Reliability, Availability and Serviceability (RAS) support for RISC-V architecture using RISC-V RERI specification and SBI SSE extension. This feature provides hardware error event handling and reporting capabilities conformant to ACPI platform error interfaces.

Key changes include:

SBI SSE Extension Support:
- Add Supervisor Software Events mechanism for non-maskable event notification from SBI firmware
- Implement context switching with firmware saving registers a6/a7
- Add __sse_entry_task per_cpu array for reliable current task tracking during SSE event delivery
- Allocate dedicated stacks for each event and CPU to support event preemption
- Handle event completion with proper interrupt simulation for signal delivery to user tasks

RAS Framework Integration:
- Integrate with existing GHES driver framework using highest priority SSE events for hardware error delivery
- Register GHES entries with SSE layer using GHES notification vectors as SSE events
- Add RISC-V specific processor type and ISA string entries
- Add GHES SSE fixmap indices for physical address mapping
- Enable build and configuration support for RAS

The implementation ensures reliable hardware error event processing and maintains system stability during error conditions. SSE events can be delivered at any time including during exception handling, with proper context preservation and signal delivery mechanisms.

Co-developed-by: Clément Léger <cleger@rivosinc.com>
Signed-off-by: Clément Léger <cleger@rivosinc.com>
Co-developed-by: Himanshu Chauhan <hchauhan@ventanamicro.com>
Signed-off-by: Himanshu Chauhan <hchauhan@ventanamicro.com>
Tested-by: lupeng <lu.peng3@zte.com.cn>
Signed-off-by: lupeng <lu.peng3@zte.com.cn>
Signed-off-by: liuqingtao <liu.qingtao2@zte.com.cn>

clementleger committed:
riscv: add SBI SSE extension definitions

community inclusion
category: feature
bugzilla: https://github.com/RVCK-Project/rvck/issues/95

------------------

Add needed definitions for SBI Supervisor Software Events extension [1].
This extension enables the SBI to inject events into supervisor software
much like ARM SDEI.

[1] https://lists.riscv.org/g/tech-prs/message/515

Signed-off-by: Clément Léger <cleger@rivosinc.com>
Link: https://patch.msgid.link/20251105082639.342973-2-cleger@rivosinc.com
Signed-off-by: Paul Walmsley <pjw@kernel.org>
Signed-off-by: Lu Peng <lu.peng3@zte.com.cn>
Signed-off-by: liuqingtao <liu.qingtao2@zte.com.cn>

GooTal commented:
cc：@yclup

oervci commented:
Kernel build success!

oervci commented:
Lava check fail! log: https://jenkins.oerv.ac.cn/job/rvck-pipeline/job/rvck-lava-trigger/153/consoleFull, result url: https://lava.oerv.ac.cn/results/447/0_rvck_common-test_qemu

clementleger committed:
riscv: add support for SBI Supervisor Software Events extension

community inclusion
category: feature
bugzilla: https://github.com/RVCK-Project/rvck/issues/95

------------------

The SBI SSE extension allows the supervisor software to be notified by
the SBI of specific events that are not maskable. The context switch is
handled partially by the firmware which will save registers a6 and a7.
When entering kernel we can rely on these 2 registers to setup the stack
and save all the registers.

Since SSE events can be delivered at any time to the kernel (including
during exception handling, we need a way to locate the current_task for
context tracking. On RISC-V, it is sotred in scratch when in user space
or tp when in kernel space (in which case SSCRATCH is zero). But at a
at the beginning of exception handling, SSCRATCH is used to swap tp and
check the origin of the exception. If interrupted at that point, then,
there is no way to reliably know were is located the current
task_struct. Even checking the interruption location won't work as SSE
event can be nested on top of each other so the original interruption
site might be lost at some point. In order to retrieve it reliably,
store the current task in an additional __sse_entry_task per_cpu array.
This array is then used to retrieve the current task based on the
hart ID that is passed to the SSE event handler in a6.

That being said, the way the current task struct is stored should
probably be reworked to find a better reliable alternative.

Since each events (and each CPU for local events) have their own
context and can preempt each other, allocate a stack (and a shadow stack
if needed for each of them (and for each cpu for local events).

When completing the event, if we were coming from kernel with interrupts
disabled, simply return there. If coming from userspace or kernel with
interrupts enabled, simulate an interrupt exception by setting IE_SIE in
CSR_IP to allow delivery of signals to user task. For instance this can
happen, when a RAS event has been generated by a user application and a
SIGBUS has been sent to a task.

Signed-off-by: Clément Léger <cleger@rivosinc.com>
Link: https://patch.msgid.link/20251105082639.342973-3-cleger@rivosinc.com
[pjw@kernel.org: cleaned up patch description and whitespace]
Signed-off-by: Paul Walmsley <pjw@kernel.org>
Signed-off-by: Lu Peng <lu.peng3@zte.com.cn>
Signed-off-by: liuqingtao <liu.qingtao2@zte.com.cn>

clementleger committed:
drivers: firmware: add riscv SSE support

community inclusion
category: feature
bugzilla: https://github.com/RVCK-Project/rvck/issues/95

--------------

Add driver level interface to use RISC-V SSE arch support. This interface
allows registering SSE handlers, and receive them. This will be used by
PMU and GHES driver.

Co-developed-by: Himanshu Chauhan <hchauhan@ventanamicro.com>
Signed-off-by: Himanshu Chauhan <hchauhan@ventanamicro.com>
Signed-off-by: Clément Léger <cleger@rivosinc.com>
Acked-by: Conor Dooley <conor.dooley@microchip.com>
Link: https://patch.msgid.link/20251105082639.342973-4-cleger@rivosinc.com
Signed-off-by: Paul Walmsley <pjw@kernel.org>
Signed-off-by: Lu Peng <lu.peng3@zte.com.cn>
Signed-off-by: liuqingtao <liu.qingtao2@zte.com.cn>

clementleger committed:
perf: RISC-V: add support for SSE event

community inclusion
category: feature
bugzilla: https://github.com/RVCK-Project/rvck/issues/95

----------------

In order to use SSE within PMU drivers, register a SSE handler for the
local PMU event. Reuse the existing overflow IRQ handler and pass
appropriate pt_regs. Add a config option RISCV_PMU_SSE to select event
delivery via SSE events.

Signed-off-by: Clément Léger <cleger@rivosinc.com>
Link: https://patch.msgid.link/20251105082639.342973-5-cleger@rivosinc.com
Signed-off-by: Paul Walmsley <pjw@kernel.org>
Signed-off-by: Lu Peng <lu.peng3@zte.com.cn>
Signed-off-by: liuqingtao <liu.qingtao2@zte.com.cn>

clementleger committed:
selftests/riscv: add SSE test module

community inclusion
category: feature
bugzilla: https://github.com/RVCK-Project/rvck/issues/95

----------------

This module, once loaded, will execute a series of tests using the SSE
framework. The provided script will check for any error reported by the
test module.

Signed-off-by: Clément Léger <cleger@rivosinc.com>
Link: https://patch.msgid.link/20251105082639.342973-6-cleger@rivosinc.com
Signed-off-by: Paul Walmsley <pjw@kernel.org>
Signed-off-by: Lu Peng <lu.peng3@zte.com.cn>
Signed-off-by: liuqingtao <liu.qingtao2@zte.com.cn>

hschauhan committed:
riscv: Define ioremap_cache for RISC-V

community inclusion
category: feature
bugzilla: https://github.com/RVCK-Project/rvck/issues/95
Link: https://lore.kernel.org/all/20250227123628.2931490-2-hchauhan@ventanamicro.com/

----------------

bert and einj drivers use ioremap_cache for mapping entries
but ioremap_cache is not defined for RISC-V.

Signed-off-by: Himanshu Chauhan <hchauhan@ventanamicro.com>
Signed-off-by: Lu Peng <lu.peng3@zte.com.cn>
Signed-off-by: liuqingtao <liu.qingtao2@zte.com.cn>

hschauhan committed:
riscv: Define arch_apei_get_mem_attribute for RISC-V

community inclusion
category: feature
bugzilla: https://github.com/RVCK-Project/rvck/issues/95
Link: https://lore.kernel.org/all/20250227123628.2931490-3-hchauhan@ventanamicro.com/

----------------

ghes_map function uses arch_apei_get_mem_attribute to get the
protection bits for a given physical address. These protection
bits are then used to map the physical address.

Signed-off-by: Himanshu Chauhan <hchauhan@ventanamicro.com>
Signed-off-by: Lu Peng <lu.peng3@zte.com.cn>
Signed-off-by: liuqingtao <liu.qingtao2@zte.com.cn>

hschauhan committed:
acpi: Introduce SSE in HEST notification types

community inclusion
category: feature
bugzilla: https://github.com/RVCK-Project/rvck/issues/95
Link: https://lore.kernel.org/all/20250227123628.2931490-4-hchauhan@ventanamicro.com/

----------------

Introduce a new HEST notification type for RISC-V SSE events.
The GHES entry's notification structure contains the notification
to be used for a given error source. For error sources delivering
events over SSE, it should contain the new SSE notification type.

Signed-off-by: Himanshu Chauhan <hchauhan@ventanamicro.com>
Signed-off-by: Lu Peng <lu.peng3@zte.com.cn>
Signed-off-by: liuqingtao <liu.qingtao2@zte.com.cn>

hschauhan committed:
riscv: Add fixmap indices for GHES IRQ and SSE contexts

community inclusion
category: feature
bugzilla: https://github.com/RVCK-Project/rvck/issues/95
Link: https://lore.kernel.org/all/20250227123628.2931490-5-hchauhan@ventanamicro.com/

----------------

GHES error handling requires fixmap entries for IRQ notifications.
Add fixmap indices for IRQ, SSE Low and High priority notifications.

Signed-off-by: Himanshu Chauhan <hchauhan@ventanamicro.com>
Signed-off-by: Lu Peng <lu.peng3@zte.com.cn>
Signed-off-by: liuqingtao <liu.qingtao2@zte.com.cn>

hschauhan committed:
riscv: conditionally compile GHES NMI spool function

community inclusion
category: feature
bugzilla: https://github.com/RVCK-Project/rvck/issues/95
Link: https://lore.kernel.org/all/20250227123628.2931490-6-hchauhan@ventanamicro.com/

----------------

Compile ghes_in_nmi_spool_from_list only when NMI and SEA
is enabled. Otherwise compilation fails with "defined but
not used" error.

Signed-off-by: Himanshu Chauhan <hchauhan@ventanamicro.com>
Signed-off-by: Lu Peng <lu.peng3@zte.com.cn>
Signed-off-by: liuqingtao <liu.qingtao2@zte.com.cn>

hschauhan committed:
riscv: Add functions to register ghes having SSE notification

community inclusion
category: feature
bugzilla: https://github.com/RVCK-Project/rvck/issues/95
Link: https://lore.kernel.org/all/20250227123628.2931490-7-hchauhan@ventanamicro.com/

----------------

Add functions to register the ghes entries which have SSE as
notification type. The vector inside the ghes is the SSE event
ID that should be registered.

Signed-off-by: Himanshu Chauhan <hchauhan@ventanamicro.com>
Signed-off-by: Lu Peng <lu.peng3@zte.com.cn>
Signed-off-by: liuqingtao <liu.qingtao2@zte.com.cn>

hschauhan committed:
riscv: Add RISC-V entries in processor type and ISA strings

community inclusion
category: feature
bugzilla: https://github.com/RVCK-Project/rvck/issues/95
Link: https://lore.kernel.org/all/20250227123628.2931490-8-hchauhan@ventanamicro.com/

----------------

- Add RISCV in processor type
- Add RISCV32/64 in ISA

Signed-off-by: Himanshu Chauhan <hchauhan@ventanamicro.com>
Signed-off-by: Lu Peng <lu.peng3@zte.com.cn>
Signed-off-by: liuqingtao <liu.qingtao2@zte.com.cn>

hschauhan committed:
riscv: Introduce HEST SSE notification handlers

community inclusion
category: feature
bugzilla: https://github.com/RVCK-Project/rvck/issues/95
Link: https://lore.kernel.org/all/20250227123628.2931490-9-hchauhan@ventanamicro.com/

----------------

- Functions to register a ghes entry with SSE
- Add Handlers for low/high priority events
- Call ghes common handler to handle an error event

Signed-off-by: Himanshu Chauhan <hchauhan@ventanamicro.com>
Signed-off-by: Lu Peng <lu.peng3@zte.com.cn>
Signed-off-by: liuqingtao <liu.qingtao2@zte.com.cn>

hschauhan committed:
riscv: Add config option to enable APEI SSE handler

community inclusion
category: feature
bugzilla: https://github.com/RVCK-Project/rvck/issues/95
Link: https://lore.kernel.org/all/20250227123628.2931490-10-hchauhan@ventanamicro.com/

----------------

APEI SSE handlers can be enabled/disabled with this config
option. When enabled, the SSE registration is done for GHES
entries having notification type set to SSE. When disabled,
registration function return not supported error.

Enable the CONFIG_ACPI_APEI and CONFIG_ACPI_APEI_GHES options in defconfig.

Signed-off-by: Himanshu Chauhan <hchauhan@ventanamicro.com>
Signed-off-by: Lu Peng <lu.peng3@zte.com.cn>
Signed-off-by: liuqingtao <liu.qingtao2@zte.com.cn>

hschauhan committed:
riscv: Enable APEI and NMI safe cmpxchg options required for RAS

community inclusion
category: feature
bugzilla: https://github.com/RVCK-Project/rvck/issues/95
Link: https://lore.kernel.org/all/20250227123628.2931490-11-hchauhan@ventanamicro.com/

----------------

Enable the APEI option so that APEI GHES options are visible.
Enable SAFE_CMPXCHG option required for GHES error handling.

Signed-off-by: Himanshu Chauhan <hchauhan@ventanamicro.com>
Signed-off-by: Lu Peng <lu.peng3@zte.com.cn>
Signed-off-by: liuqingtao <liu.qingtao2@zte.com.cn>

oervci commented:
check patch done. log: https://jenkins.oerv.ac.cn/job/rvck-pipeline/job/check-patch/132/consoleFull

oervci commented:
Kernel build success!

wangliu-iscas commented:
/check

oervci commented:
check patch done. log: https://jenkins.oerv.ac.cn/job/rvck-pipeline/job/check-patch/138/consoleFull

oervci commented:
Kernel build success!

sterling-teng commented:
开发树已经滚动至6.6.103，请尽快更新并rebase。

GooTal commented:
> 开发树已经滚动至6.6.103，请尽快更新并rebase。

已更新，烦请审阅

oervci commented:
kunit test done. log:https://jenkins.oerv.ac.cn/job/rvck-pipeline/job/kunit-test/167/consoleFull

oervci commented:
check patch done. log: https://jenkins.oerv.ac.cn/job/rvck-pipeline/job/check-patch/162/consoleFull

oervci commented:
Kernel build failed! log:https://jenkins.oerv.ac.cn/job/rvck-pipeline/job/rvck-kernel-build/194/consoleFull

GooTal commented:
@oervci 开始测试

oervci commented:
kunit test done. log:https://jenkins.oerv.ac.cn/job/rvck-pipeline/job/kunit-test/169/consoleFull

oervci commented:
开始测试

oervci commented:
Kernel build success!

oervci commented:
kunit test done. log:https://jenkins.oerv.ac.cn/job/rvck-pipeline/job/kunit-test/179/consoleFull

oervci commented:
Lava check done! result url: https://lava.oerv.ac.cn/results/736/0_rvck_common-test_qemu

oervci commented:
check patch done. log: https://jenkins.oerv.ac.cn/job/rvck-pipeline/job/check-patch/177/consoleFull

oervci commented:
kunit test done. log:https://jenkins.oerv.ac.cn/job/rvck-pipeline/job/kunit-test/182/consoleFull

oervci commented:
Kernel build success!

oervci commented:
kunit test done. log:https://jenkins.oerv.ac.cn/job/rvck-pipeline/job/kunit-test/184/consoleFull

oervci commented:
Lava check done! result url: https://lava.oerv.ac.cn/results/747/0_rvck_common-test_qemu

sterling-teng commented:
defconfig配置验证完了吗？

sterling-teng commented:
“ edk2和opensbi的配套RAS功能已提出pr，”现在那边是合并了，还是在审阅中？
